### PR TITLE
[Network] Azure Bastion native client support for az cli

### PR DESCRIPTION
**Pull request by aavalang**
**Description**
AZ cli support for native client ssh and rdp through tunneling of Azure Bastions.

**Testing Guide**
az network bastion ssh --resource-group <Rg> --name <name> --vm <vm-name> --auth-type <auth-type>
az network bastion rdp --resource-group <Rg> --name <name> --vm <vm-name>

**History Notes**

[Network] az network bastion ssh: connect to a Virtual machine through ssh using Bastion Tunneling.
[Network] az network bastion rdp: connect to a Virtual machine through native RDP using Bastion Tunneling.
[Network] az network bastion tunnel: connect to a Virtual machine using Bastion Tunneling.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 1]
```diff
@@ -7044,6 +7044,36 @@
...
+helps['network bastion tunnel'] = """
+type: command
+short-summary: Show a Azure bastion host machine.
+examples:
+  - name: Show a Azure bastion host machine (autogenerated)
+    text: |
+        az network bastion tunnel --name MyBastionHost --resource-group MyResourceGroup --target-resource-id vmResourceId --resource-port 111 --port 222
+    crafted: true
+"""
+
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:7076]: Please remove `(autogenerated)` and `crafted: true` for manual examples.

```suggestion
helps['network bastion ssh'] = """
type: command
short-summary: SSH to a virtual machine using Tunneling from Azure Bastion.
examples:
  - name: SSH to virtual machine using Azure Bastion
    text: |
        az network bastion ssh --name MyBastionHost --resource-group MyResourceGroup --target-resource-id vmResourceId --auth-type AAD/password/ssh-key-file --username xyz
"""

helps['network bastion rdp'] = """
type: command
short-summary: RDP to target Virtual Machine using Tunneling from Azure Bastion.
examples:
  - name: RDP to virtual machine using Azure Bastion
    text: |
        az network bastion rdp --name MyBastionHost --resource-group MyResourceGroup --target-resource-id vmResourceId
"""

helps['network bastion tunnel'] = """
type: command
short-summary: Show a Azure bastion host machine.
examples:
  - name: Show a Azure bastion host machine
    text: |
        az network bastion tunnel --name MyBastionHost --resource-group MyResourceGroup --target-resource-id vmResourceId --resource-port 111 --port 222
"""

```

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 2]
```diff
@@ -7044,6 +7044,36 @@
...
 """
 
+helps['network bastion ssh'] = """
+type: command
+short-summary: SSH to a virtual machine using Tunneling from Azure Bastion.
+examples:
+  - name: SSH to virtual machine using Azure Bastion (autogenerated)
+    text: |
+        az network bastion ssh --name MyBastionHost --resource-group MyResourceGroup --target-resource-id vmResourceId --auth-type AAD/password/ssh-key-file --username xyz
+    crafted: true
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:7054]: Please add different examples for different `--auth-type`

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -15,12 +15,20 @@
...
+    UnrecognizedArgumentError, ResourceNotFoundError
+from azure.cli.core.profiles import ResourceType, supported_api_version
 
+from azure.cli.command_modules.network._client_factory import network_client_factory
 from azure.cli.command_modules.network.zone_file.parse_zone_file import parse_zone_file
 from azure.cli.command_modules.network.zone_file.make_zone_file import make_zone_file
-from azure.cli.core.profiles import ResourceType, supported_api_version
-from azure.cli.core.azclierror import ResourceNotFoundError, UnrecognizedArgumentError
+
+from .tunnel import TunnelServer
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:26]: Could you move this import line into the function used?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 4]
```diff
@@ -7451,6 +7459,174 @@ def list_bastion_host(cmd, resource_group_name=None):
...
+    else:
+        raise UnrecognizedArgumentError("Unknown auth type. Use one of password, aad or akv.")
+    command = command + ["-p", str(tunnel_server.local_port)]
+    command = command + ['-o', "StrictHostKeyChecking=no", '-o', "UserKnownHostsFile=/dev/null"]
+    command = command + ['-o', "LogLevel=Error"]
+    logger.debug("Running ssh command %s", ' '.join(command))
+    try:
+        subprocess.call(command, shell=platform.system() == 'Windows')
+    except Exception as ex:
+        raise CLIError(ex)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:7582]: `CLIError` is deprecated, please use other kinds of error instead.

**Review by kairu-ms — CHANGES_REQUESTED** [review]
There's no tests for these commands. So I still recommand you move them into extensions.

File: src/azure-cli/azure/cli/command_modules/network/_help.py (hunk 1 above)
Comment (aavalang) [reply src/azure-cli/azure/cli/command_modules/network/_help.py]: fixed

File: src/azure-cli/azure/cli/command_modules/network/_help.py (hunk 2 above)
Comment (aavalang) [reply src/azure-cli/azure/cli/command_modules/network/_help.py]: added

**Comment by aavalang** [comment]
> There's no tests for these commands. So I still recommand you move them into extensions.

unable to mock tunnel as discussed

**Review by msabansal — CHANGES_REQUESTED** [review]

**Review by msabansal — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/network/tunnel.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            else:
+                logger.info('Port %s is open', self.local_port)
+                is_port_open = True
+            return is_port_open
+
+    def _get_auth_token(self):
+        profile = Profile(cli_ctx=self.cli_ctx)
+        user = profile.get_current_account_user()
+        # Generate an Azure token with the VSTS resource app id
+        auth_token = profile.get_access_token_for_resource(user, None, None)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/network/tunnel.py:79]: `get_access_token_for_resource` will be deprecated pretty soon (https://github.com/Azure/azure-cli/issues/18944). Please migrate to `get_raw_token`:

https://github.com/Azure/azure-cli/blob/14cc787d0f58bc649d402b486fdecc5625eee9ac/src/azure-cli-core/azure/cli/core/_profile.py#L679

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/network/tunnel.py:79]: Why is the third argument `resource` set as `None`?

https://github.com/Azure/azure-cli/blob/14cc787d0f58bc649d402b486fdecc5625eee9ac/src/azure-cli-core/azure/cli/core/_profile.py#L531-L535

AAD will default it to `00000002-0000-0000-c000-000000000000` which is AD Graph API.

File: src/azure-cli/azure/cli/command_modules/network/tunnel.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                                        sockopt=((socket.IPPROTO_TCP, socket.TCP_NODELAY, 1),),
+                                        sslopt={'cert_reqs': ssl.CERT_NONE},
+                                        enable_multithread=True)
+            logger.info('Websocket, connected status: %s', self.ws.connected)
+            index = index + 1
+            logger.info('Got debugger connection... index: %s', index)
+            debugger_thread = Thread(target=self._listen_to_client, args=(self.client, self.ws, index))
+            web_socket_thread = Thread(target=self._listen_to_web_socket, args=(self.client, self.ws, index))
+            debugger_thread.start()
+            web_socket_thread.start()
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/tunnel.py:130]: App Service have update the implementation of tunnel.py please update it.

https://github.com/Azure/azure-cli/blob/213997916f90666e0e70050b24ca94c17fc4ab55/src/azure-cli/azure/cli/command_modules/appservice/tunnel.py#L139-L153

File: src/azure-cli/azure/cli/command_modules/network/tunnel.py (hunk 5 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/network/tunnel.py:79]: The only other place where `get_access_token_for_resource` is called is `vsts_cd_provider.py`:

https://github.com/Azure/azure-cli/blob/0a22ec50affe656cd95a210890beb860425523ee/src/azure-cli/azure/cli/command_modules/appservice/vsts_cd_provider.py#L33

`vsts_cd_provider.py` has been removed by #18203:

```
> git log -- src/azure-cli/azure/cli/command_modules/appservice/vsts_cd_provider.py
commit 8701d602677a6e26f3ab681c2ee590879ff2d5c7
Author: Calvin <calvinch4n@gmail.com>
Date:   Thu Jul 1 18:10:37 2021 -0700

    [AppService] BREAKING CHANGE: az webapp deployment source config: Remove vsts-cd-manager (#18203)

    * Remove vsts-cd-manager

    * Rerun tests

    * Remove vsts-cd-manager

    * Rerun tests
```

**Pull request merged**